components/brave_wallet_ui/components/desktop/hardware-wallet-connect/hardware_wallet_connect.tsx

```diff
@@ -117,205 +168,195 @@ export const HardwareWalletConnect = ({
+      return (
+        scheme.coin === selectedAccountType.coin &&
+        scheme.vendor === selectedHardwareVendor
+      )
+    })
+  }, [selectedHardwareVendor, selectedAccountType.coin])
+
+  const loadMoreAccounts = React.useCallback(
+    async (
+      startIndex: number,
+      numberOfAccountsToLoad: number,
+      scheme: HardwareImportScheme
+    ) => {
+      setIsLoadingAccounts(true)
+      const result = await loadAccountsFromDevice({
+        startIndex,
+        count: numberOfAccountsToLoad,
+        scheme,
         onAuthorized: hideAuthorizeDevice
       })
-        .then((result) => {
-          setAccounts(result)
-        })
-        .catch((error) => {
-          setConnectionError(getErrorMessage(error, selectedAccountType.name))
-          setShowAccountsList(false)
-        })
-        .finally(() => setIsConnecting(false))
-    },
-    [selectedHardwareWallet, selectedAccountType, filecoinNetwork]
-  )
+      setIsLoadingAccounts(false)
+
+      if (result.payload) {
+        setShowAccountsList(true)
 
-  const getDefaultAccountName = React.useCallback(
-    (account: BraveWallet.HardwareWalletAccount) => {
-      const index = accounts.findIndex((e) => e.address === account.address)
-      let schemeString
-      switch (selectedDerivationScheme) {
-        case LedgerDerivationPaths.Legacy:
-          schemeString = ' (Legacy)'
-          break
-        default:
-          schemeString = ''
+        const newAccounts: AccountFromDeviceListItem[] = result.payload.map(
+          (acc) => {
+            const alreadyInWallet = !!savedAccounts.find((a) => {
+              return (
+                a.accountId.kind === BraveWallet.AccountKind.kHardware &&
+                a.address === acc.address
+              )
+            })
+            return { ...acc, alreadyInWallet, shouldAddToWallet: false }
+          }
+        )
+
+        setAccounts((prev) => prev.concat(newAccounts))
+        return
       }
 
-      return index === 0
-        ? `${account.hardwareVendor}${schemeString}`
-        : `${account.hardwareVendor} ${index}${schemeString}`
+      setShowAccountsList(false)
+      if (result.error === 'unauthorized') {
+        setShowAuthorizeDevice(true)
+      } else {
+        setConnectionError(
+          getErrorMessage(result.error, selectedAccountType.name)
+        )
+      }
     },
-    [accounts, selectedDerivationScheme]
+    [savedAccounts, selectedAccountType.name]
   )
 
-  const onAddAccounts = React.useCallback(() => {
-    const selectedAccounts = accounts.filter((account) =>
-      selectedDerivationPaths.includes(account.derivationPath)
+  React.useEffect(() => {
+    // Don't load more if already loading.
+    if (isLoadingAccounts) {
+      return
+    }
+
+    // Just return if there is nothing to load.
+    const numberOfAccountsToLoad = totalNumberOfAccounts - accounts.length
+    if (numberOfAccountsToLoad <= 0) {
+      return
+    }
+
+    loadMoreAccounts(
```

Comment:
we should inline the `loadMoreAccounts` logic here so we can better handle race-conditions.

You'll want to add an `ignore` variable to check before setting any state and return a "cleanup" function from the effect

https://react.dev/reference/react/useEffect#fetching-data-with-effects